src/Components/ProtectedRoute.js

```diff
@@ -0,0 +1,20 @@
+import { Navigate } from "react-router-dom";
+import { useContext } from "react";
+import { AuthenticationContext } from "../Contexts/AuthenticationContenxt";
+
+function ProtectedRoute({ component, unprotected }) {
+  const { user } = useContext(AuthenticationContext);
+
+  if (unprotected) {
+    if (user) {
+      return <Navigate to="/notes" />;
+    }
+    return component;
+  }
+  if (!user) {
+    return <Navigate to="/signin" />;
+  }
+  return component;
+}
```

Comment:
you can put the `Route` component here, in that way you don't even need to use the vanilla Route component in your App.js, you simply call this special protectedRoute component instead of the standard one.

Reply:
Tried it and got this error:
`Uncaught Error: [ProtectedRoute] is not a <Route> component. All component children of <Routes> must be a <Route> or <React.Fragment>`

Event with this simplified version:
**App.js:**
```javascript
import { Routes } from "react-router-dom";
import ProtectedRoute from "./Components/ProtectedRoute";

import Layout from "./Components/Layout";

function App() {
  return (
    <div className="App">
      <Routes>
        <ProtectedRoute path="/" element={<Layout />} />
      </Routes>
    </div>
  );
}

export default App;
```

**ProtectedRoute.js**
```javascript
import { Route } from "react-router-dom";

function ProtectedRoute() {
  return <Route path="/" element="Lorem ipsum..." />;
}

export default ProtectedRoute;
```

An explanation I found on StackOverflow: 
_In RRDv6 the Route components don't have a render prop, and only Route or React.Fragment are valid children of a Routes component. ProtectedRoute is neither of these and will fail an invariant check._

Reply:
Ah it seems that this changed since the last time I used router. Here is a good explaination how to solve this:
https://www.robinwieruch.de/react-router-private-routes/
You basically wrap all protected routes in one `Route` component with the element set to your ProtectedRoute component. This Wrapper Route then has all the Routes that shall be protected as its direct children, like this: 
```js
<Route element={<ProtectedRoute user={user} />}>
  <Route path="home" element={<Home />} />
  <Route path="dashboard" element={<Dashboard />} />
</Route>
```

Reply:
You need to use the `Outlet` component from react router in your ProtectedRoute though.